docs/design-notes/reverse-reads.md

```diff
@@ -0,0 +1,123 @@
+# Reverse reads
+
+A read is called reverse when it reads with reverse clustering order
+(compared to that of the schema). Example:
+
+    CREATE TABLE mytable (
+        pk int,
+        ck int,
+        s int STATIC,
+        v int,
+        PRIMARY KEY (pk, ck)
+    ) WITH
+        CLUSTERING ORDER BY (ck ASC);
+
+    # Forward read (using table's native order)
+    SELECT * FROM mytable WHERE pk = 1;
+    # Explicit forward order
+    SELECT * FROM mytable WHERE pk = 1 ORDER BY ck ASC;
+
+    # Reverse read
+    SELECT * FROM mytable WHERE pk = 1 ORDER BY ck DESC;
+
+If the table's native clustering order is DESC, then a read with ASC
+order is considered reverse.
+
+## Legacy format
+
+The legacy format is how scylla handled reverse queries internally. We
+are in the process of migrating to the native reverse format, but for
+now coordinator-side code still uses the legacy format.
+
+### Request
+
+The `query::partition_slice::options::reversed` flag is set.
+Clustering ranges in both `query::partition_slice::_row_ranges` and
+`query::specific_ranges::_ranges`
+(`query::partition_slice::_specific_ranges`) are half-reversed: they
+are ordered in reverse, but when they are compared to other
+mutation-fragments, their end bound is used as position, instead of the
+start bound as usual. When compared to other clustering ranges the end
+bound is used as the start bound and vice-versa.
+Example:
+
+For the clustering keys (ASC order): `ck1`, `ck2`, `ck3`, `ck4`, `ck5`,
+`ck6`.
+A `_row_ranges` field of a slice might contain this:
+
+    [ck1, ck2], [ck4, ck5]
+
+The legacy reversed version would look like this:
+
+    [ck4, ck5], [ck1, ck2]
+
+Note how the ranges themselves are the same (bounds not reversed), it is
+just the range vector itself that is reversed.
+
+### Result
+
+Results are ordered with the reversed clustering order with the caveat
+that range-tombstones are ordered by their end bound, using the native
+schema's comparators. For example given the following partition:
+
+    ps{pk1}, sr{}, cr{ck1}, rt{[ck2, ck4)}, cr{ck2}, cr{ck3}, cr{ck4}, ck{ck5}, pe{}
+
+The legacy reverse format equivalent of this looks like the following:
+
+    ps{pk1}, sr{}, cr{ck5}, rt{[ck2, ck4)}, cr{ck4}, cr{ck3}, cr{ck2}, ck{ck1}, pe{}
```

Comment:
So the reversed stream does not conform to the reversed schema. This is contrary to what we decided at the beginning. Why?

Reply:
This is the existing "legacy" format. For the native one, to which what you said applies see below.

Reply:
Oh, this is about reconcilable_result, not mutation fragment stream?

Reply:
Both use this format actually.